rfcs/20241205-graph-api-causal-mask/README.md

```diff
@@ -0,0 +1,485 @@
+which is crucial when dealing with large-scale models and extended sequence lengths.
+
+## Support in frameworks and libraries
+
+### cuDNN
+
+cuDNN supports both explicit and implicit causal masks in its frontend and backend
+APIs. In the frontend API, users can set SDPA attributes for various mask types,
+including arbitrary user-generated mask, causal mask, padding mask, alibi mask and
+sliding window mask on an SDPA node [3]. The backend API requires users to construct
+a graph using operations like `GenIndex`, `Greater Than` and `Select` to indicate
+a causal mask [4].
+
+### PyTorch
+
+PyTorch supports both explicit and implicit causal masks in the aten operator
+`scaled_dot_product_attention` [5]:
+
+```
+Tensor scaled_dot_product_attention(
+    const Tensor& query_,
+    const Tensor& key,
+    const Tensor& value,
+    const std::optional<Tensor>& attn_mask_,
+    double dropout_p,
+    bool is_causal,
+    std::optional<double> scale)
+```
+
+Existing models can incorporate causal masks in two main ways:
+
+#### 1. Manual implementation of SDPA with small operations
+
+Some models use small operations to generate SDPA functionality. For example,
+in the BERT definition in torchbenchmark [6], the SDPA forward is defined as:
+
+```python
+scores = torch.matmul(query, key.transpose(-2, -1)) / math.sqrt(query.size(-1))
+if mask is not None:
+    min_mask = -1e9
+    scores = scores.masked_fill(mask == 0, min_mask)
+p_attn = F.softmax(scores, dim=-1)
+return torch.matmul(p_attn, value), p_attn
+```
+
+These operations are passed to `torch.compile` and fused into the aten operator
+of `scaled_dot_product_attention`. In this case, the mask is explicitly generated
+in the user code, and the aten operator must accept it.
+
+#### 2. Direct use of SDPA operator
+
+Users can also employ the aten operator directly, using
+`torch.nn.functional.scaled_dot_product_attention` [7], which is later lowered to
+the aten operator for computation. This approach offers two options:
+
+##### a. Explicit mask
+
+For example, in gpt-fast [8]:
+
+```python
+y = F.scaled_dot_product_attention(q, k, v, attn_mask=mask, dropout_p=0.0)
+```
+
+This method is similar to the manual implementation approach, using an explicit mask.
+
+##### b. Implicit mask
+
+As seen in the nano_gpt implementation in torchbenchmark [9]:
+
+```python
+y = F.scaled_dot_product_attention(q, k, v, attn_mask=None, dropout_p=0.0, is_causal=True)
+```
+
+In this case, the aten operator utilizes different SDPA backends, each handling
+the is_causal boolean flag differently, for example:
+
+- cuDNN attention backend uses the cuDNN frontend API with this flag set.
+- Flash attention backend's CPU implementation masks unnecessary GEMM output with
+negative infinity [10].
+
+If integrating oneDNN Graph into the aten operation for this scenario, the
+integration currently needs to generate a causal mask inside the aten operator, and
+pass it to the library as an explicit mask.
```

Comment:
Hi @dzarukin , thank you for the review.

> It feels that generating explicit mask on a framework side would kill the benefit of memory savings which is the major treasure hunt here...

Yes, exactly. But there are two user scenarios as mentioned in [PyTorch](https://github.com/ElaineBao/oneDNN/blob/graph-api-causal-mask-support/rfcs/20241205-graph-api-causal-mask/README.md#pytorch) part. 
1. If user manually implements SDPA with small operations, then the mask is already there, integrating oneDNN Graph doesn't need to generate a mask additionally.
2. If user directly uses the fused operation - SDPA, there are also two cases:
    a. using it with an explicit mask: then similar to the scenario 1, no need to generate a mask additionally.
    b. using the fused operation with an implicit mask: this is the case that framework has to generate a mask for oneDNN Graph.

In summary, 1. and 2.a. are the cases that user chooses to give up the memory savings by themselves' design consideration, while 2.b. is the case that user has to give up the memory savings because of the current support status of oneDNN Graph. This RFC mainly focuses on solving 2.b.


> Does the request this RFC covers intends to support aten operation?

Yes, since in most user scenarios, the aten operation will be used.